core/src/nucleus/mod.rs

```diff
@@ -104,77 +171,208 @@ impl FunctionResult {
     }
 }
 
+
+/// Enum of all Actions that mutates the Nucleus's state
 #[derive(Clone, Debug, PartialEq)]
 pub enum Action {
     InitApplication(Dna),
+    ReturnInitializationResult(bool),
     ExecuteZomeFunction(FunctionCall),
     ReturnZomeFunctionResult(FunctionResult),
+    ValidateEntry(EntrySubmission),
+}
+
+
+/// Reduce ReturnInitializationResult Action
+/// On initialization success, set Initialized status
+/// otherwise reset the nucleus
+fn reduce_rir(nucleus_state : & mut NucleusState, has_succeeded : bool) {
+    assert!(nucleus_state.status == NucleusStatus::Initializing);
+    if has_succeeded {
+        (*nucleus_state).status = NucleusStatus::Initialized
+    } else {
+        *nucleus_state = NucleusState::new();
```

Comment:
I'm not sure I totally understand what you were thinking here, but it seems like if the initialization didn't succeed we can't just reset the state, to uninitialized, because a client of the state object actually might want to know that the initialization failed and distinguish between these states.

Reply:
My thought was that if init fails, then undo everything that was done in the InitApplication Action reducer, which basically means resetting the state. In the absence of clarity for the whole App init procedure (#86) it seemed like the safest thing to do for now. I agree about needing some callback to notify that the init failed. I am not satisfied with the wait loop on history counts.